### PR TITLE
catalog: Add E2E migration tests

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds end-to-end tests using mzcompose that test the
catalog implementation migration and rollback processes.

Works towards resolving #24218

### Motivation
Adds tests

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Comment by jkosh44** [comment]
@def- This is my WIP for adding mzcompose tests for catalog migrations. My 3 open questions are:

1. What's left of the current test that is not commented out is just a sanity check, it should be passing. However, when I run it locally it fails. Any idea what's going on? It seems like I'm not stopping and the starting the services correctly to preserve the data.
2. I couldn't figure out how to format my `workflow_default` method such that it ran all the other workflows.
3. I wasn't entirely sure how to add this test to CI.

**Comment by def-** [comment]
I just fixed them up. I hope my commit makes it clear, otherwise I can explain:

> What's left of the current test that is not commented out is just a sanity check, it should be passing. However, when I run it locally it fails. Any idea what's going on? It seems like I'm not stopping and the starting the services correctly to preserve the data.

Every Testdrive without no_reset=True was resetting the state.

> I couldn't figure out how to format my workflow_default method such that it ran all the other workflows.

You had the right idea in the commented out code, but had to exclude `default`, otherwise endless recursion.

> I wasn't entirely sure how to add this test to CI.

test/ci/pipeline.template.yaml

File: test/persist-catalog-migration/mzcompose.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+from materialize.mzcompose.composition import Composition
+from materialize.mzcompose.services.cockroach import Cockroach
+from materialize.mzcompose.services.materialized import Materialized
+from materialize.mzcompose.services.minio import Minio
+from materialize.mzcompose.services.testdrive import Testdrive
+
+SERVICES = [
+    Minio(setup_materialize=True),
+    Cockroach(setup_materialize=True),
```
Comment (def-) [inline test/persist-catalog-migration/mzcompose.py:20]: You don't seem to use external_minio and external_cockroach, so I think these services are not required.

File: test/persist-catalog-migration/mzcompose.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    c.up("minio", "cockroach")
+    c.up("testdrive", persistent=True)
+
+    # Switch to emergency stash catalog so only the stash epoch is incremented.
+    c.up("materialized")
+    c.testdrive(
+        input=dedent(
+            """
+    $ postgres-execute connection=postgres://mz_system@materialized:6877/materialize
+    ALTER SYSTEM SET catalog_kind TO 'emergency-stash'
```
Comment (def-) [inline test/persist-catalog-migration/mzcompose.py:90]: Would it be more realistic to set this during startup of materialized already? You could use `c.override(Materialized(...)):` for that.

**Review by def- — APPROVED** [review]

File: test/persist-catalog-migration/mzcompose.py (hunk 1 above)
Comment (jkosh44) [reply test/persist-catalog-migration/mzcompose.py]: Thanks, removed.

File: test/persist-catalog-migration/mzcompose.py (hunk 2 above)
Comment (jkosh44) [reply test/persist-catalog-migration/mzcompose.py]: Yeah, just updated to do this.

**Pull request merged**